### PR TITLE
Bluetooth: Host: Fix issue where more than `CONFIG_BT_EATT_MAX` EATT channels may be created

### LINKED ISSUE #42306: Bluetooth: Host: More than `CONFIG_BT_EATT_MAX` EATT channels may be created
**Describe the bug**
It is possible to connect more than `CONFIG_BT_EATT_MAX` enhanced ATT bearers on a connection if `CONFIG_BT_MAX_CONN` > 1.

The check on [att.c:2896-2905](https://github.com/zephyrproject-rtos/zephyr/blob/b374dc6a7fd8f841394ad6cf2364c7ceef05f3c7/subsys/bluetooth/host/att.c#L2896-L2905) does not work when a credit based connection request has been sent and not responded to yet, because the `bt_att_chan` objects have not yet been added to the `bt_att::chans` list. Allocation from `chan_slab` succeeds if `CONFIG_BT_MAX_CONN` > 1, even though `CONFIG_BT_EATT_MAX` channels have been allocated for the given connection.

**To Reproduce**
Run the babblesim test in #42268 with `CONFIG_BT_MAX_CONN=2`. 10 EATT channels will be connected, the test expects 5.
1. Add `CONFIG_BT_MAX_CONN=2` to tests/bluetooth/bsim_bt/bsim_test_eatt/prj_collision.conf
2. Run tests/bluetooth/bsim_bt/compile.sh
3. Run tests/bluetooth/bsim_bt/bsim_test_eatt/tests_scripts/collision.sh
4. See error "Expected 5 enhanced bearers, got 10"

**Expected behavior**
I would expect `att_chan_new()` to fail (ie. return `NULL`) if `CONFIG_BT_EATT_MAX` enhanced ATT channels are already allocated for the given connection.

**Impact**
Additional connections may become unusable if a credit based connection collision happens.

### PR DESCRIPTION
**Pull request by hermabe**
Fixes #42306.

This PR depends on #42268 as it uses the same test procedure to reproduce the bug.

### CODE REVIEW

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
The commit "Fix check for number of ATT bearers" looks good to me. But can you rebase this branch to not not include the changes going into #42268? Otherwise, we will need to wait with this PR until the other is merged. It's ok to commit the same test code on both branches. It will merge/rebase nicely later.

**Review by alwa-nordic — DISMISSED** [review]

**Comment by hermabe** [comment]
FYI: The test fails because a credit based connection collision happens and the mitigation is not in this PR.

**Comment by carlescufi** [comment]
Please do not use GitHub issue numbers in the commit titles, be descriptive instead.

**Comment by carlescufi** [comment]
@hermabe why is this still Draft? Can you also rebase it please?

**Comment by hermabe** [comment]
It depends on the L2CAP collision mitigation PR (#42268), will rebase once that one is merged. It is just missing the final approval by @jhedberg, so should be merged soon/once 3.0 is released.

**Comment by hermabe** [comment]
Rebased on main. 

`tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/provision/pb_adv_reprovision.sh` fails. Bisect gives df9b8721e2 as first bad commit, but I am not sure why it makes the test fail.

**Comment by hermabe** [comment]
CI is green and the PR is ready to review now.

**Review by jhedberg — DISMISSED** [review]

**Review by alwa-nordic — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/att.c
Comment (hermabe) [reply subsys/bluetooth/host/att.c:2590]: Wouldn't `chan` also always be non-null? Could this whole function be replaced with `CONTAINTER_OF(conn, struct bt_att, conn)`?

**Review by Thalley — CHANGES_REQUESTED** [review]

**Review by Thalley — DISMISSED** [review]

File: subsys/bluetooth/host/att.c
Comment (alwa-nordic) [reply subsys/bluetooth/host/att.c:2590]: > Wouldn't chan also always be non-null? Could this whole function be replaced with CONTAINTER_OF(conn, struct bt_att, conn)?

No. The `CONTAINER_OF` operation "unfocuses" an address to get the surrounding structure (in the same object). `bt_conn` is not part of the `bt_att_chan` structure. `bt_att_chan` just holds a pointer to a `bt_conn`.

File: subsys/bluetooth/host/att.c
Comment (alwa-nordic) [reply subsys/bluetooth/host/att.c:2725]: This is related to the removal of the line below. That was the single user of this value.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
Comments that need resolution:
[assert instead of BT_WARN](https://github.com/zephyrproject-rtos/zephyr/pull/42359/files#r813849901)
[unused value](https://github.com/zephyrproject-rtos/zephyr/pull/42359/files#r819477506)

**Comment by hermabe** [comment]
> Comments that need resolution:
[assert instead of BT_WARN](https://github.com/zephyrproject-rtos/zephyr/pull/42359/files#r813849901)
[unused value](https://github.com/zephyrproject-rtos/zephyr/pull/42359/files#r819477506)

I think these are resolved now. Will rebase on main to fix CI.

**Comment by hermabe** [comment]
Bluetooth tests are failing, but seems unrelated to this change.
> fatal: invalid upstream 'origin/main'

**Comment by hermabe** [comment]
Will have to wait for https://github.com/zephyrproject-rtos/zephyr/pull/43426 to be merged before re-running the CI.

**Review by alwa-nordic — DISMISSED** [review]

**Review by Thalley — APPROVED** [review]

**Comment by hermabe** [comment]
@jhedberg @alwa-nordic Is there anything missing or can we merge this now?

**Review by jhedberg — APPROVED** [review]

**Comment by hermabe** [comment]
@carlescufi This is no longer DNM and can be merged.

**Pull request merged**
